sdk/src/azure/platform/az_pahoasync.c

```diff
@@ -0,0 +1,619 @@
+      out_prop_strpair->key = az_span_create((uint8_t*)prop->value.data.data, prop->value.data.len);
+      out_prop_strpair->value
+          = az_span_create((uint8_t*)prop->value.value.data, prop->value.value.len);
+      return AZ_OK;
+    }
+    stringpair_index++;
+    prop = MQTTProperties_getPropertyAt(
+        property_bag->pahoasync_properties, (enum MQTTPropertyCodes)type, stringpair_index);
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+AZ_NODISCARD az_result az_mqtt5_property_bag_read_byte(
+    az_mqtt5_property_bag* property_bag,
+    az_mqtt5_property_type type,
+    uint8_t* out_prop_byte)
+{
+  _az_PRECONDITION_NOT_NULL(property_bag);
+  _az_PRECONDITION_NOT_NULL(out_prop_byte);
+  MQTTProperty* prop;
+
+  prop = MQTTProperties_getProperty(
+      property_bag->pahoasync_properties, (enum MQTTPropertyCodes)type);
+
+  if (prop == NULL)
+  {
+    return AZ_ERROR_ITEM_NOT_FOUND;
+  }
+
+  *out_prop_byte = prop->value.byte;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_mqtt5_property_bag_read_int(
+    az_mqtt5_property_bag* property_bag,
+    az_mqtt5_property_type type,
+    uint32_t* out_prop_int)
+{
+  _az_PRECONDITION_NOT_NULL(property_bag);
+  _az_PRECONDITION_NOT_NULL(out_prop_int);
+  MQTTProperty* prop;
+
+  prop = MQTTProperties_getProperty(
+      property_bag->pahoasync_properties, (enum MQTTPropertyCodes)type);
+
+  if (prop == NULL)
+  {
+    return AZ_ERROR_ITEM_NOT_FOUND;
+  }
+
+  // TODO_L: Add reading capabilities for integer 2.
+  *out_prop_int = prop->value.integer4;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_mqtt5_property_bag_read_binarydata(
+    az_mqtt5_property_bag* property_bag,
+    az_mqtt5_property_type type,
+    az_mqtt5_property_binarydata* out_prop_bindata)
+{
+  _az_PRECONDITION_NOT_NULL(property_bag);
+  _az_PRECONDITION_NOT_NULL(out_prop_bindata);
+  MQTTProperty* prop;
+
+  prop = MQTTProperties_getProperty(
+      property_bag->pahoasync_properties, (enum MQTTPropertyCodes)type);
+
+  if (prop == NULL)
+  {
+    return AZ_ERROR_ITEM_NOT_FOUND;
+  }
+
+  out_prop_bindata->bindata = az_span_create((uint8_t*)prop->value.data.data, prop->value.data.len);
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_span az_mqtt5_property_get_string(az_mqtt5_property_string* prop_str)
+{
+  _az_PRECONDITION_NOT_NULL(prop_str);
+
+  return prop_str->str;
+}
+
+AZ_NODISCARD az_span
+az_mqtt5_property_stringpair_get_key(az_mqtt5_property_stringpair* prop_strpair)
+{
+  _az_PRECONDITION_NOT_NULL(prop_strpair);
+
+  return prop_strpair->key;
+}
+
+AZ_NODISCARD az_span
+az_mqtt5_property_stringpair_get_value(az_mqtt5_property_stringpair* prop_strpair)
+{
+  _az_PRECONDITION_NOT_NULL(prop_strpair);
+
+  return prop_strpair->value;
+}
+
+AZ_NODISCARD az_span az_mqtt5_property_get_binarydata(az_mqtt5_property_binarydata* prop_bindata)
+{
+  _az_PRECONDITION_NOT_NULL(prop_bindata);
+
+  return prop_bindata->bindata;
+}
+
+void az_mqtt5_property_free_string(az_mqtt5_property_string* prop_str) { (void)prop_str; }
```

Comment:
Does the application need to free this themselves because they allocate it? If so, would it make sense to still provide the free in here or not?
Or is it that nothing gets allocated so nothing needs to be free'd?

Reply:
Nothing gets allocated, and if they do allocate a MQTTProperty they would need to free it themselves, not us since Paho itself doesn't allocate those except when it receives a message